### ISSUE TITLE
feature: add folding to LSP

### ISSUE THREAD

**Issue by jackielii** [a-h/templ#650]
I believe the LSP can add folding support. I believe this is very useful when you have long html tags. Folding them can make one read the code more easily.

**Comment by a-h** [comment]
From the LSP spec, it looks like it is possible to add that. https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_foldingRange

Personally, I never use folding in editors.

The change would be applied here: https://github.com/a-h/templ/blob/ad6950151f9c0a33ff0c3123ac8128ded4d0dfe2/cmd/templ/lspcmd/proxy/server.go#L716-L722

First, you'd need to get the source code of the current templ file from the in-memory cache, as per https://github.com/a-h/templ/blob/ad6950151f9c0a33ff0c3123ac8128ded4d0dfe2/cmd/templ/lspcmd/proxy/server.go#L728-L735

Then, you'd build on @alehechka's recent work in https://github.com/a-h/templ/commit/22f761a91ef666b76bee251269bd81219259b0c0, where he added tracking the positions of HTML elements in the parser tree.

With those changes, it's now possible to walk through the templ object model, find the locations of things you want to fold (e.g. HTML elements etc) and return back the list of folding ranges in the LSP. The structure of a templ file is laid out in https://github.com/a-h/templ/blob/main/parser/v2/structure.go - once parsed, those are the nodes you're looking for in the doc.

This won't be high on the priority list for me, but I'd be happy to take a PR for the change. Anyone reading, let me know if you're going to try.

If you want to build this feature, and aren't sure how LSP stuff works, I did a talk at Gophercon UK about it, and I talk about how it works in templ, so do check it out: https://www.youtube.com/watch?v=EkK8Jxjj95s

**Comment by jackielii** [comment]
@a-h Reading the links you provided, I can't help but feeling this would be covered by the "html intellisense" that's proposed here: https://github.com/a-h/templ/issues/498

@alehechka what do you think? Is it worth adding just the folding or is it better to have the full HTML capabilities here that should include folding?

**Comment by joerdav** [comment]
I believe that this won't come by html intellisense being integrated. We will want do our own folding since we will want to allow folding of html, go code and custom templ code. So proxying the html lsp won't suffice here.

**Comment by sirgallifrey** [comment]
I'll play around with this, but I don't have experience working with LSPs, so no promises

**Comment by sirgallifrey** [comment]
@a-h I was able to make good progress, so I'm now confident I'll have a PR soon.

 still have to write all the tests, and figure out if I'll be able to add folding to Go code, but the rest is looking good.
 
 
![Screen Recording 2024-04-07 at 02 45 56](https://github.com/a-h/templ/assets/2331645/910562e2-db25-4d39-af2a-d01d8aa0ce65)

**Comment by joerdav** [comment]
Great work! Hopefully go folding will be a case of proxying gopls

**Comment by sirgallifrey** [comment]
Unfortunately that gif I posted was a blunter, turned out my templ build was not working and that is the default folding VSCode does when there's no LSP

(I I'll not delete my comment to perpetuate my shame 😄 )

But I'm not too far off. My only issue is that the way I did is eating the last line (see pictures)
The spec says you can send a line number and a col number. The col number by default is the end of the line, and col number is optional and even if the serves sends them, the client can choose to ignore.

so my issue is that I send the line number of the closing bracket, on the column right before it and seems VS code is just ignoring it and consuming the entire line.

![Screenshot 2024-04-08 at 22 06 24](https://github.com/a-h/templ/assets/2331645/6e41103f-9272-4ec7-ab5b-7a4c9fe3c9d8)
![Screenshot 2024-04-08 at 22 06 41](https://github.com/a-h/templ/assets/2331645/212f08b5-77d3-441b-a940-ac351201c29e)

You can see the folded shows only `<head>` while the default VSCode implementation would also show the closing tag. Same thing for the brackets.

**Comment by sirgallifrey** [comment]
@joerdav Yes, That's what I'll try to do, I guess i just have to make sure I just pic the foldings of gopls that correspond to Templs TemplateFileGoExpression nodes

**Comment by joerdav** [comment]
Related to this, for treesitter users folds can be defined by treesitter. I've got a PR for this open, and can confirm it works with nvim.

https://github.com/vrischmann/tree-sitter-templ/pull/43

**Comment by joerdav** [comment]
@sirgallifrey Any more progress on this? I'm wondering if the default folding from vscode is enough?

**Comment by sirgallifrey** [comment]
Hello @joerdav,

Good news/Bad news

Bad news:
I did implemented folding on the Server, but it didn't worked well because I implemented the folding ranges using start/end characters, but it seems VSCode supports only full line folding ranges.

Full line folding ranges get a little more annoying to implement because they don't quite match with the parsing as it is now, would require some back tracking... nothing hard... just something that don't fit very well and there's where I lost my motivation to work on it. 
Here is how far I got, in case someone want to pick it up and enhance it: https://github.com/a-h/templ/pull/810/files

Good news:
I figured out how to use the default folding of VSCode that seems to work very well, There was no direct way to do it, but I extended the LanguageClient class and prevented it from ever loading the Folding feature. Sounds like a hack, but not so bad. I'll create a PR shortly on templ-vscode
